[FE] ErrorBoundary & Suspense를 임시로 적용한다

airman5573 opened the pull request:
## 요약
몇몇 컴포넌트에 ErrorBoundary와 Suspense를 임시로 도입했습니다.  
이런 스타일로 진행해도 괜찮다면, 나머지 컴포넌트에도 적용해보겠습니다.  

## 세부사항
### useSuspendedQuery
react-query에서는 suspense옵션을 활성화 해도 return하는 data는 undefined일 수 있다고 합니다.  
[react-query issue](https://github.com/TanStack/query/issues/1297)  
(1) enabled option과 data의 type을 연결짓는일이 어렵고
(2) 도중에 query를 cancel했을때도 핸들링하기가 어렵다고 합니다.  

그래서 enabled는 항상 true + cancel는 안한다는 전제 하에 useSuspendedQuery를 만들었습니다.  
엄밀히는 [toss/slash](https://slash.page/libraries/react/react-query/src/hooks/usesuspendedquery.i18n/)에서 코드를 가져왔습니다.  

### react-error-bundary
React에서 제공하는 ErrorBoundary대신 [react-error-boundary](https://github.com/bvaughn/react-error-boundary)를 사용했습니다.  
react-query와 궁합이 잘맞는다고 생각하고, 사용하기도 편합니다.  
이를테면, `<ErrorBoundary FallbackComponent={MyErrorFallback}>...</ErrorBoundary>` 이렇게 쓰면  
MyErrorFallback은 error객체를 props로 전달받습니다. 즉, 편합니다!  
또한 [react-query와 함께 reset을 handling하는 경우에도 더 편합니다](https://react-query-v3.tanstack.com/guides/suspense).

verus-j committed:
Merge pull request #442 from woowacourse-teams/develop

[BE] 스프린트 6

jaejae-yoo committed:
Merge pull request #443 from woowacourse-teams/develop

fix: 불필요한 폴더 제거

jaejae-yoo committed:
docs: 인프라 구조 수정

tco0427 committed:
Merge pull request #451 from woowacourse-teams/develop

스프린트6 Merge

airman5573 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2022-moamoa into develop

airman5573 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2022-moamoa into develop

airman5573 committed:
feat: useSuspendedQuery 추가

airman5573 committed:
feat: MainPage에 errorboundary & suspense 추가

airman5573 committed:
feat: DetailPage에 errorboundary & suspense를 적용한다

nan-noo commented on frontend/src/hooks/useSuspendedQuery.ts:
```diff
@@ -0,0 +1,83 @@
+// from: https://github.com/toss/slash/blob/main/packages/react/react-query/src/hooks/useSuspendedQuery.ts
+import { QueryFunction, QueryKey, UseQueryOptions, UseQueryResult, useQuery } from 'react-query';
```
type을 붙여줍시다!

airman5573 replied:
수정했습니다 :D
[05de07e](https://github.com/woowacourse-teams/2022-moamoa/pull/476/commits/05de07e224a06560cb82aad218cc3c2da1a8aae5)

nan-noo commented on frontend/src/hooks/useSuspendedQuery.ts:
```diff
@@ -0,0 +1,83 @@
+  extends Omit<UseQueryResult<TData, never>, 'data' | 'status' | 'error' | 'isLoading' | 'isError' | 'isFetching'> {
+  data: TData;
+  status: 'success' | 'idle';
+}
```
interface가 쓰였네요!

airman5573 replied:
수정했습니다 :D
[30b8639e](https://github.com/woowacourse-teams/2022-moamoa/pull/476/commits/30b8639eea78ee119a1849132c04a00c32fade60)

nan-noo commented on frontend/src/hooks/useSuspendedQuery.ts:
```diff
@@ -0,0 +1,83 @@
+  queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+  options: SuspendedUseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+): SuspendedUseQueryResultOnSuccess<TData> | SuspendedUseQueryResultOnIdle<undefined>;
+export function useSuspendedQuery<
```
올려준 이슈 자료 봤습니다
https://github.com/TanStack/query/issues/1297#issuecomment-730459493

요거 때문이군요!

airman5573 replied:
맞습니다!

nan-noo commented on frontend/src/hooks/useSuspendedQuery.ts:
```diff
@@ -0,0 +1,83 @@
+  isIdle: true;
+};
+
+export type SuspendedUseQueryOptions<
```
통일성을 위해서
UseSuspendedQueryOptions는 어떤가요?

nan-noo commented on frontend/src/hooks/useSuspendedQuery.ts:
```diff
@@ -0,0 +1,83 @@
+  isSuccess: true;
+  isIdle: false;
+};
+export type SuspendedUseQueryResultOnIdle<TData> = BaseSuspendedUseQueryResult<TData> & {
```
이 타입들도 UseSuspendedQuery~ 가 어떤가요?

airman5573 replied:
수정했습니다 :D  
[662f6b0](https://github.com/woowacourse-teams/2022-moamoa/pull/476/commits/662f6b0643a393927f19ded68a27c111cfde93a7)

nan-noo commented on frontend/src/hooks/useSuspendedQuery.ts:
```diff
@@ -0,0 +1,83 @@
+  TQueryKey extends QueryKey = QueryKey,
+> = Omit<UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>, 'suspense'>;
+
+export function useSuspendedQuery<
+  TQueryFnData = unknown,
+  TError = unknown,
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey,
+>(
+  queryKey: TQueryKey,
+  queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+  options?: Omit<SuspendedUseQueryOptions<TQueryFnData, TError, TData, TQueryKey>, 'enabled'>,
+): SuspendedUseQueryResultOnSuccess<TData>;
+export function useSuspendedQuery<
+  TQueryFnData = unknown,
+  TError = unknown,
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey,
+>(
+  queryKey: TQueryKey,
+  queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+  options: Omit<SuspendedUseQueryOptions<TQueryFnData, TError, TData, TQueryKey>, 'enabled'> & {
+    enabled?: true;
+  },
+): SuspendedUseQueryResultOnSuccess<TData>;
```
요 두가지 경우가
- enabled를 안받는 경우
- enabled가 undefined(optional)이거나 true인 경우

이고 return 타입은 같은데, 두번째 경우가 첫번째 경우를 포함하는 것 아닌가요??
enabled를 생략하면 optional에 걸리니까요!

airman5573 replied:
앞에 options가 optional이 달라서 두개가 필요한듯 싶습니다.  

1. options를 넣어도 되고 안넣어도 되는데 -> 어쨋든 enabled 는 옵션 항목에서 제공하지 않을것이다(안받을것이다)
2. options를 꼭 넣어야 겠다면 -> enabled는 넣지 말거나 true로 해라!

nan-noo replied:
아 옵셔널이 있었군요

nan-noo commented on frontend/src/hooks/useSuspendedQuery.ts:
```diff
@@ -0,0 +1,83 @@
+  queryFn: QueryFunction<TQueryFnData, TQueryKey>,
+  options?: SuspendedUseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+) {
+  return useQuery(queryKey, queryFn, { ...options, suspense: true }) as BaseSuspendedUseQueryResult<TData>;
```
as를 빼면 어떤 경고가 나오나요?

airman5573 replied:
경고는 안나오지만 BaseSuspendedUseQueryResult 타입으로 캐스팅 하는 편이 useSuspendedQuery 훅의 결과값으로 더 잘 어울린다고 생각합니다.  

```typescript
interface BaseSuspendedUseQueryResult<TData>
  extends Omit<UseQueryResult<TData, never>, 'data' | 'status' | 'error' | 'isLoading' | 'isError' | 'isFetching'> {
  data: TData;
  status: 'success' | 'idle';
}
```
suspense를 사용하고 있기 때문에 'error' | 'isLoading' | 'isError' | 'isFetching' 가 필요 없으니까요!

nan-noo commented on frontend/src/hooks/useSuspendedQuery.ts:
```diff
@@ -0,0 +1,83 @@
+// from: https://github.com/toss/slash/blob/main/packages/react/react-query/src/hooks/useSuspendedQuery.ts
```
관련 이슈 및 설명 주석을 달면 좋을 것 같아요! 오버로드된 함수가 많으니까
useSuspendedQuery
- suspense: true
- enabled 옵션이 optional이거나 true면 원본 데이터를 리턴
- enabled 옵션이 false면 데이터가 undefined
- 그 외의 경우 원본 데이터 | undefined

airman5573 replied:
설명 작성했습니다 :D  
[eb705f4](https://github.com/woowacourse-teams/2022-moamoa/pull/476/commits/eb705f47d108dc20d06b47f9d9dc17710a5734af)

nan-noo commented on frontend/src/api/studies/index.ts:
```diff
@@ -77,6 +77,7 @@ export const useGetInfiniteStudies = ({ title, selectedFilters }: ApiInfiniteStu
         if (!lastPage.hasNext) return;
         return { page: lastPage.page };
       },
+      suspense: true,
```
아 이건 인피니트쿼리라 훅을 사용하지 못했군요

nan-noo replied:
프로젝트에서 인피니트 쿼리를 단순 쿼리만큼 많이 사용하는데, 혹시 그 슬래시에서 서스펜스 인피니트 쿼리 훅은 없었나요..? ㅋㅋㅋ
비슷하게 만들면 될 것 같긴한데, 타입 오버로드하는게 번거롭네용

airman5573 replied:
저도 같은 생각인데, 혹시 태태 함 도전해 보실래요?

nan-noo commented on frontend/package.json:
```diff
@@ -80,6 +80,7 @@
     "marked": "^4.0.18",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
+    "react-error-boundary": "^3.1.4",
```
https://tanstack.com/query/v4/docs/guides/suspense#resetting-error-boundaries

여기서 저도 보긴 했는데(스타도 많고 버전도 높고 공식문서에서도 나오고), 아직 제가 에러 바운더리를 직접 구현해본 적이 없어서 라이브러리 말고 직접 구현하고 싶습니다 :D

airman5573 replied:
코드 [요기](https://github.com/bvaughn/react-error-boundary/blob/master/src/index.tsx)있습니다.  
크게 어려워 보이는 코드는 아니네요!

태태가 다음 PR에 만드시는건가요?

nan-noo replied:
아마 만든다면 다다음 PR일 것 같네요ㅎㅎ

요거 머지하고 스켈레톤 하고 그다음에 갈아끼우는 방식으로..!

nan-noo commented on frontend/src/api/study/index.ts:
```diff
@@ -8,6 +8,8 @@ import type { MakeOptional, StudyDetail, StudyId, TagId } from '@custom-types';
 import axiosInstance from '@api/axiosInstance';
 import { checkStudy } from '@api/study/typeChecker';
 
+import { useSuspendedQuery } from '@hooks/useSuspendedQuery';
```
👍

nan-noo commented on frontend/src/pages/detail-page/DetailPage.tsx:
```diff
@@ -30,71 +32,80 @@ const DetailPage: React.FC = () => {
   return (
     <PageWrapper>
-      {(() => {
-        if (isFetching) return <Loading />;
-        if (isError) return <Error />;
-        if (isSuccess)
```
확실히 깔끔해지네요! 역할 분리가 잘 된 느낌이 확 들어요

nan-noo replied:
아 근데 원래 로딩과 에러 부분이 PageWrapper 내부에 있었는데 외부로 이동하면서 스타일이 바뀔 것 같아요!(스켈레톤까지 적용하고 나면 많이 바뀔 수도 있어 보여요)
서스펜스랑 에러바운더리를 내부로 옮기거나 페이지 랩퍼를 외부로 옮겨야 할 것 같습니다!

airman5573 replied:
맞네요!

nan-noo commented on frontend/src/pages/detail-page/DetailPage.tsx:
```diff
@@ -30,71 +32,80 @@ const DetailPage: React.FC = () => {
     </PageWrapper>
   );
 };
+const DetailPage: React.FC = () => {
+  return (
+    <ErrorBoundary FallbackComponent={ErrorFallback}>
+      <Suspense fallback={<LoadingFallback />}>
+        <Core />
+      </Suspense>
+    </ErrorBoundary>
+  );
+};
```
여기로 페이지 랩퍼를 옮기면 좋을 것 같네요!

nan-noo commented on frontend/src/pages/detail-page/DetailPage.tsx:
```diff
@@ -30,71 +32,80 @@ const DetailPage: React.FC = () => {
+  return (
+    <ErrorBoundary FallbackComponent={ErrorFallback}>
+      <Suspense fallback={<LoadingFallback />}>
+        <Core />
```
코어 대신 원래 사용하던 Self나 Content는 어떤가요? Page도 있고..!
그리고 이 DetailPage 컴포넌트가 맨 위에 있으면 좋을 것 같아요 :D

nan-noo replied:
음 아니면 원래 DetailPage를 그대로 두고 페이지를 서스펜스랑 에러바운더리로 감싼 어떤 컴포넌트를 만들어도 될 것 같네요 나중에 반복된다 싶으면

airman5573 replied:
네이밍이 쉽지 않더라구요  
Self, Content, Page는 중복이 발생할것 같아요.  
'성공하는 경우만 다루는 메인 컨탠츠' 이런 느낌인데 ~  
Main 어떤가요?

nan-noo replied:
한국어 그대로 해서 MainContent 좋을 것 같은데요!

airman5573 replied:
유선상으로 회의한대로 "Success"로 변경했습니다 :D  
[afb1746](https://github.com/woowacourse-teams/2022-moamoa/pull/476/commits/afb1746d155a949855ea77296b361879fbcd2ad9)

nan-noo commented on frontend/src/pages/main-page/MainPage.tsx:
```diff
@@ -45,11 +46,34 @@ const MainPage: React.FC = () => {
 };
 
+const MainPage: React.FC = () => {
+  return (
```
이것도 DetailPage와 마찬가지로 PageWrapper가 여기 있어야 할 것 같아요!

airman5573 replied:
Loading, ErrorBoundary 쪽에 각각 PageWrapper를 두는건 어떨까요?  
```typescript
const ErrorFallback: React.ComponentType<FallbackProps> = ({ error }) => {
  return (
    <PageWrapper>
      <h2>스터디를 불러오는 도중 에러가 발생했습니다</h2>
      <p>{error.message}</p>
    </PageWrapper>
  );
};

const LoadingFallback: React.FC = () => {
  return (
    <PageWrapper>
      <div>스터디 불러오는중</div>
    </PageWrapper>
  );
};
```

왜냐하면 
```typescript
return (
    <>
      <FilterSection selectedFilters={selectedFilters} onFilterButtonClick={handleFilterButtonClick} />
      <CreateNewStudyButton onClick={handleCreateNewStudyButtonClick} />
      <PageWrapper>
        {(() => {
          if (!searchedStudies || (searchedStudies && searchedStudies.length === 0)) return <NoResult />;
          return (
            <InfinitScrollCardList
              isContentLoading={isFetching}
              onContentLoad={fetchNextPage}
              studies={searchedStudies}
            />
          );
        })()}
      </PageWrapper>
    </>
  );
```

이런 경우가 있기 때문입니다.

nan-noo commented on frontend/src/pages/main-page/MainPage.tsx:
```diff
@@ -75,7 +99,7 @@ const InfinitScrollCardList: React.FC<InfinitScrollCardListProps> = ({
       ))}
     </CardList>
-    {isContentLoading && <Loading />}
+    {isContentLoading && <LoadingFallback />}
```
만약 로딩이 발생하면 서스펜스에 걸리나요 isFetching이 실행되나요? 제 생각엔 suspense true로 하면 서스펜스에 우선 걸려서 이 부분이 렌더링되지 않을 것 같은데 어떻게 생각하시나요?
서스펜스를 사용하려면 이 부분에 해야 할 것 같습니다!

\+ 수정이 많이 필요할 것 같으면 어차피 제가 스켈레톤 구현하면서 이 메인페이지 로딩 로직을 수정해서 크게 바꿀 필요는 없습니다 :D

airman5573 replied:
infiniteQuery는 처음에만 suspense에 걸리고 그 다음에는 안걸리더라구요.  
아마 로딩될때마다 서스펜스에 걸리면 ui가 너무 안좋아져서 그런거 아닐까 싶어요.

그나저나 태태! 다운받아서 한번 돌려봐 주세요! 🤣

nan-noo reviewed (changes requested):
리뷰 코멘트에 남겼습니다 :D 
useSuspendedQuery 훅 좋습니다

airman5573 committed:
refactor: type을 import하는 경우 type을 붙여준다

airman5573 committed:
refactor: interface대신 type을 사용한다

- extends대신 Merge를 통해 두개의 타입을 병합한다. & 으로도 가능하지만, 이 경우 A & B 에서 A에 있는 요소와 B에있는 요소가 동시에 반영된다. B가 오버라이드 하지 않는것이다.

airman5573 committed:
refactor: SuspendedUseQuery.. -> UseSuspendedQuery..

airman5573 committed:
refactor: useSuspendedQuery 훅의 등장 배경을 기록한다

airman5573 committed:
refactor: 로딩에 성공한 경우 보여주는 View의 이름 변경(Core -> Success)

airman5573 committed:
feat: Error/Loading View를 Wrapper로 감싼다
